system_metrics_collector/src/topic_statistics_collector/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,373 @@
+#include "libstatistics_collector/topic_statistics_collector/topic_statistics_collector.hpp"
+#include "system_metrics_collector/constants.hpp"
+#include "system_metrics_collector/metrics_message_publisher.hpp"
+
+
+namespace topic_statistics_collector
+{
+
+using libstatistics_collector::topic_statistics_collector::TopicStatisticsCollector;
+using libstatistics_collector::topic_statistics_collector::ReceivedMessageAgeCollector;
+using libstatistics_collector::topic_statistics_collector::ReceivedMessagePeriodCollector;
+namespace constants =
+  libstatistics_collector::topic_statistics_collector::topic_statistics_constants;
+
+/**
+ * Build a positive integer range to use in parameter descriptors.
+ *
+ * @return a positive integer range with specififies start, end and step values
+ */
+inline rcl_interfaces::msg::IntegerRange buildPositiveIntegerRange(
+  int64_t from,
+  int64_t to,
+  uint64_t step)
+{
+  rcl_interfaces::msg::IntegerRange positive_range;
+  positive_range.from_value = from;
+  positive_range.to_value = to;
+  positive_range.step = step;
+
+  return positive_range;
+}
+
+/**
+ * Build a parameter description for period node parameters.
+ *
+ * @return a read-only integer range topic descriptor
+ */
+inline rcl_interfaces::msg::ParameterDescriptor buildPeriodParameterDescriptor(
+  const std::string & description)
+{
+  const auto positive_range = buildPositiveIntegerRange(
+    1,
+    std::numeric_limits<decltype(rcl_interfaces::msg::IntegerRange::to_value)>::max(),
+    1
+  );
+
+  rcl_interfaces::msg::ParameterDescriptor period_descriptor;
+  period_descriptor.read_only = true;
+  period_descriptor.integer_range.push_back(positive_range);
+  period_descriptor.description = description;
+
+  return period_descriptor;
+}
+
+/**
+ * Build a parameter description for string node parameters.
+ *
+ * @return a read-only topic descriptor
+ */
+inline rcl_interfaces::msg::ParameterDescriptor buildTopicParameterDescriptor(
+  const std::string & description)
+{
+  rcl_interfaces::msg::ParameterDescriptor topic_descriptor;
+  topic_descriptor.read_only = true;
+
+  return topic_descriptor;
+}
+
+/**
+ * Validate the values assigned to string node parameters.
+ *
+ * @throws std::invalid_argument if the parameter value is empty
+ */
+inline void validateStringParam(const std::string & name, const std::string & value)
+{
+  rcpputils::require_true(!value.empty(), name + " node parameter cannot be empty");
+}
+
+/**
+ * Initialize topic statistics collectors to use with this node.
+ */
+template<typename T>
+void initializeCollectors(
+  std::vector<std::unique_ptr<TopicStatisticsCollector<T>>> & container)
+{
+  container.push_back(std::make_unique<ReceivedMessageAgeCollector<T>>());
+  container.push_back(std::make_unique<ReceivedMessagePeriodCollector<T>>());
+}
+
+/**
+ * Class which makes periodic topic statistics measurements, using a ROS2 timer.
+ */
+template<typename T>
```

Comment:
Please document what is T here. Can we add a `static_assert` to make sure only "allowed types" are passed?

Reply:
`T` here is for any ROS 2 message type. I'm not sure how we can enforce that using `static_assert`?